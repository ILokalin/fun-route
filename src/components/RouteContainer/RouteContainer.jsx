import React from 'react';
import PropTypes from 'prop-types';
import RoutePoint from '../RoutePoint';



export default class extends React.PureComponent {
  constructor (props) {
    super(props);

    this.handlePoint = this.handlePoint.bind(this);
    this.newIndexReady = this.newIndexReady.bind(this);
  }


  static propTypes = {
    routePoints:    PropTypes.array.isRequired,
    onDeletePoint:  PropTypes.func.isRequired,
    onViewPoint:    PropTypes.func.isRequired,
    onChangeSequence: PropTypes.func.isRequired
  }


  /**
   * Управление событиями на точке маршрута:
   * - нажатие кнопок "вверх"/"вниз"
   * - нажатие кнопки "удалить"
   * - нажатие на элемент для раскрытия подробного содержания и показа точки в центер карты
   * @function
   * @name handlePoint
   * @params {object.event} event - событие
   */
  handlePoint (event) {
    const target = event.target;    
    const itemIndex = target.closest('.route__item') ? target.closest('.route__item').dataset.index : ''

    const button = target.closest('.route__item-button');

    if (typeof(button) !== 'undefined' && button) {
      if (button.value === 'up') {
        const position = this.positionIdInRoute(itemIndex);
        const newIndex = (position === 0) ? document.querySelectorAll('.route__item').length -1 : position - 1;
        this.props.onChangeSequence(newIndex, itemIndex);

      } else if (button.value === 'down') {
        const position = this.positionIdInRoute(itemIndex);
        const newIndex = (position === document.querySelectorAll('.route__item').length -1) ? 0 : position + 1;
        this.props.onChangeSequence(newIndex, itemIndex);

      } else if (button.value === 'remove') {
        this.props.onDeletePoint(itemIndex)

      }
    } else {

      const caption = target.closest('.route__item')
      
      if (typeof(caption) !== 'undefined' && caption)  {

        if (!caption.classList.contains('route__item--active')) {
          const deactivateList = [...caption.closest('.route').querySelectorAll('.route__item')]

          deactivateList.forEach(elem => {
          elem.classList.remove('route__item--active')
          })

          this.props.onViewPoint(itemIndex)
        }

        caption.closest('.route__item').classList.toggle('route__item--active')
      }
    }
  }


  /**
   * Поиск позиции в спике маршрута по id
   * Применяется для смены последовательсности точек
   * @function
   * @name positionIdInRoute
   * @params {string} id - уникальный идентификатор
   * @return {number} - текущий индекс элемента в списке маршрута
   */
  positionIdInRoute (id) {
    const itemsList = [...document.querySelector('.route').children]
    return itemsList.findIndex(element => element.dataset.index === id);
  }


  /**
   * После перемещения точки в маршруте, при готовности нового индекса, 
   * вызов запроса на смену последовательности в соновном наборе данных
   */
  newIndexReady (id) {
    const newIndex = this.positionIdInRoute(id)
    this.props.onChangeSequence(newIndex, id);
  }


  render() {
    const {routePoints} = this.props;

    return (
      <ul className="route" onClick ={this.handlePoint}>
        {routePoints.map((point, index) =>  <RoutePoint   
                                              point = {point}
                                              key   = {point.geometry.id} 
                                              pointLetter   = {index < 26 ? String.fromCharCode(index + 65) : ''}
                                              newIndexFind  = {this.newIndexReady}
                                              />)}
      </ul>
    );
  }
}