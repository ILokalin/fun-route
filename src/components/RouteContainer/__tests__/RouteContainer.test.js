import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import rendrer from 'react-test-renderer';

import RouteContainer from 'components/RouteContainer';


describe('RouteContainer Component', () => {
  const routePoints = [{
    name: 'fake name for test',
    geometry: {
      id: 'fake_id_00000',
      getCoordinates: () => {
        return [55,37]
      }
    },
    properties: {
      get: (param) => {
        if (param === 'balloonContent') {
          return 'fake address for test 0001';
        } else {
          return 'ERROR! Check parameters "point.properties.get("balloonContent")" for address';
        }
      }
    }
  },
  {
    name: 'fake name for test',
    geometry: {
      id: 'fake_id_00001',
      getCoordinates: () => {
        return [58,58]
      }
    },
    properties: {
      get: (param) => {
        if (param === 'balloonContent') {
          return 'fake address for test 0002';
        } else {
          return 'ERROR! Check parameters "point.properties.get("balloonContent")" for address';
        }
      }
    }
  }]

  it('render snap', () => {
    const component = rendrer.create(<RouteContainer  routePoints={routePoints} 
                                                      />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    })


})