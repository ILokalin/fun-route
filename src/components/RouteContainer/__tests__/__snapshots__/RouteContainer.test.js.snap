// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteContainer Component render snap 1`] = `
<ul
  className="route"
  onClick={[Function]}
>
  <li
    className="route__item"
    data-index="fake_id_00000"
    draggable="true"
    onDragStart={[Function]}
  >
    <div
      className="route__item-control"
    >
      <h3
        className="route__item-letter"
      >
        A
      </h3>
      <button
        className="route__item-button"
        type="button"
        value="up"
      >
        <svg
          className="icon"
          fill="none"
          viewBox="0 0 20 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <title
            id="title"
          >
            Переместить точку
          </title>
          <path
            className="icon__path"
            d="M10 0c.297 0 .593.113.82.339l8.84 8.84a1.158 1.158 0 11-1.637 1.639L10 2.796l-8.023 8.022A1.158 1.158 0 11.34 9.18L9.181.339C9.407.113 9.703 0 10 0z"
            fill="#798995"
          />
        </svg>
      </button>
      <button
        className="route__item-button"
        type="button"
        value="down"
      >
        <svg
          className="icon"
          fill="none"
          viewBox="0 0 20 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <title
            id="title"
          >
            Переместить точку
          </title>
          <path
            className="icon__path"
            d="M10 11.158c-.297 0-.593-.113-.82-.339L.34 1.978A1.158 1.158 0 111.976.34L10 8.362 18.023.34a1.158 1.158 0 111.638 1.638l-8.842 8.841a1.154 1.154 0 01-.819.339z"
            fill="#798995"
          />
        </svg>
      </button>
    </div>
    <div
      className="route__item-caption"
      title="координаты: 55°00′00″  37°00′00″"
    >
      <h3
        className="route__item-address"
      >
        fake address for test 0001
      </h3>
      <p
        className="route__item-name"
      >
        fake name for test
      </p>
    </div>
    <button
      className="route__item-button route__item-button--remove"
      type="button"
      value="remove"
    >
      <svg
        className="icon"
        fill="none"
        viewBox="0 0 30 30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs />
        <title
          id="title"
        >
          Убрать точку маршрута
        </title>
        <g
          className="icon__path"
          clipPath="url(#clip0)"
          fill="#798995"
        >
          <path
            d="M19.143 10.857c.458.457.458 1.2 0 1.657L16.657 15l2.486 2.486a1.172 1.172 0 11-1.657 1.657L15 16.657l-2.486 2.486a1.172 1.172 0 01-1.657-1.657L13.343 15l-2.486-2.486a1.172 1.172 0 111.657-1.657L15 13.343l2.486-2.486a1.172 1.172 0 011.657 0zM20.602 3.621l-1.516-.628a1.172 1.172 0 01.897-2.165l1.515.628a1.172 1.172 0 11-.896 2.165z"
          />
          <path
            d="M15 0a1.172 1.172 0 110 2.344C8.005 2.344 2.344 8.004 2.344 15c0 6.995 5.66 12.656 12.656 12.656 6.995 0 12.656-5.66 12.656-12.656A1.172 1.172 0 0130 15c0 8.29-6.709 15-15 15-8.29 0-15-6.709-15-15C0 6.71 6.709 0 15 0zM24.53 6.63l-1.16-1.16a1.172 1.172 0 111.656-1.657l1.16 1.16a1.172 1.172 0 11-1.657 1.658zM28.538 11.549a1.172 1.172 0 01-1.531-.634l-.628-1.516a1.172 1.172 0 112.165-.897l.628 1.516a1.172 1.172 0 01-.634 1.53z"
          />
        </g>
        <defs>
          <clipPath
            id="clip0"
          >
            <path
              d="M0 0h30v30H0z"
              fill="#fff"
              transform="matrix(-1 0 0 1 30 0)"
            />
          </clipPath>
        </defs>
      </svg>
    </button>
  </li>
  <li
    className="route__item"
    data-index="fake_id_00001"
    draggable="true"
    onDragStart={[Function]}
  >
    <div
      className="route__item-control"
    >
      <h3
        className="route__item-letter"
      >
        B
      </h3>
      <button
        className="route__item-button"
        type="button"
        value="up"
      >
        <svg
          className="icon"
          fill="none"
          viewBox="0 0 20 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <title
            id="title"
          >
            Переместить точку
          </title>
          <path
            className="icon__path"
            d="M10 0c.297 0 .593.113.82.339l8.84 8.84a1.158 1.158 0 11-1.637 1.639L10 2.796l-8.023 8.022A1.158 1.158 0 11.34 9.18L9.181.339C9.407.113 9.703 0 10 0z"
            fill="#798995"
          />
        </svg>
      </button>
      <button
        className="route__item-button"
        type="button"
        value="down"
      >
        <svg
          className="icon"
          fill="none"
          viewBox="0 0 20 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <title
            id="title"
          >
            Переместить точку
          </title>
          <path
            className="icon__path"
            d="M10 11.158c-.297 0-.593-.113-.82-.339L.34 1.978A1.158 1.158 0 111.976.34L10 8.362 18.023.34a1.158 1.158 0 111.638 1.638l-8.842 8.841a1.154 1.154 0 01-.819.339z"
            fill="#798995"
          />
        </svg>
      </button>
    </div>
    <div
      className="route__item-caption"
      title="координаты: 58°00′00″  58°00′00″"
    >
      <h3
        className="route__item-address"
      >
        fake address for test 0002
      </h3>
      <p
        className="route__item-name"
      >
        fake name for test
      </p>
    </div>
    <button
      className="route__item-button route__item-button--remove"
      type="button"
      value="remove"
    >
      <svg
        className="icon"
        fill="none"
        viewBox="0 0 30 30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs />
        <title
          id="title"
        >
          Убрать точку маршрута
        </title>
        <g
          className="icon__path"
          clipPath="url(#clip0)"
          fill="#798995"
        >
          <path
            d="M19.143 10.857c.458.457.458 1.2 0 1.657L16.657 15l2.486 2.486a1.172 1.172 0 11-1.657 1.657L15 16.657l-2.486 2.486a1.172 1.172 0 01-1.657-1.657L13.343 15l-2.486-2.486a1.172 1.172 0 111.657-1.657L15 13.343l2.486-2.486a1.172 1.172 0 011.657 0zM20.602 3.621l-1.516-.628a1.172 1.172 0 01.897-2.165l1.515.628a1.172 1.172 0 11-.896 2.165z"
          />
          <path
            d="M15 0a1.172 1.172 0 110 2.344C8.005 2.344 2.344 8.004 2.344 15c0 6.995 5.66 12.656 12.656 12.656 6.995 0 12.656-5.66 12.656-12.656A1.172 1.172 0 0130 15c0 8.29-6.709 15-15 15-8.29 0-15-6.709-15-15C0 6.71 6.709 0 15 0zM24.53 6.63l-1.16-1.16a1.172 1.172 0 111.656-1.657l1.16 1.16a1.172 1.172 0 11-1.657 1.658zM28.538 11.549a1.172 1.172 0 01-1.531-.634l-.628-1.516a1.172 1.172 0 112.165-.897l.628 1.516a1.172 1.172 0 01-.634 1.53z"
          />
        </g>
        <defs>
          <clipPath
            id="clip0"
          >
            <path
              d="M0 0h30v30H0z"
              fill="#fff"
              transform="matrix(-1 0 0 1 30 0)"
            />
          </clipPath>
        </defs>
      </svg>
    </button>
  </li>
</ul>
`;
